source/extensions/filters/network/ratelimit/ratelimit.h

```diff
@@ -83,13 +85,44 @@ class Filter : public Network::ReadFilter,
   void initializeReadFilterCallbacks(Network::ReadFilterCallbacks& callbacks) override {
     filter_callbacks_ = &callbacks;
     filter_callbacks_->connection().addConnectionCallbacks(*this);
+    applySubstitutionFormatter(config_->descriptors(),
+                               filter_callbacks_->connection().streamInfo());
+  }
+
+  void applySubstitutionFormatter(std::vector<RateLimit::Descriptor> original_descriptors,
+                                  StreamInfo::StreamInfo& stream_info) {
+
+    std::vector<RateLimit::Descriptor> dynamicDescriptors = std::vector<RateLimit::Descriptor>();
+    for (const RateLimit::Descriptor& descriptor : original_descriptors) {
+      RateLimit::Descriptor new_descriptor;
+      for (const RateLimit::DescriptorEntry& descriptorEntry : descriptor.entries_) {
+        std::string value = descriptorEntry.value_;
+        value = formatValue(value, stream_info);
+        new_descriptor.entries_.push_back({descriptorEntry.key_, value});
+      }
+      dynamicDescriptors.push_back(new_descriptor);
+    }
+    filter_descriptors_ = dynamicDescriptors;
+  }
+
+  std::string formatValue(std::string body, StreamInfo::StreamInfo& stream_info) {
+    Http::RequestHeaderMapPtr request_headers = Http::RequestHeaderMapImpl::create();
+    Http::ResponseHeaderMapPtr response_headers = Http::ResponseHeaderMapImpl::create();
+    Http::ResponseTrailerMapPtr response_trailers = Http::ResponseTrailerMapImpl::create();
+
+    Formatter::FormatterImpl formatter(body);
```

Comment:
The way this is being done is fairly inefficient. You are rebuilding the formatter every time you want to apply a format.  The way to do this is to build the formatter at configuration load time, and then you can apply it for each connection. So essentially, the formatter build code should live in the config object, and then you can consume it from the filter.